ASAN_TRAP | WTF::HashTable::lookup; WebCore::LegacyRenderSVGResourceClipper::removeClientFromCache; WebCore::SVGResources::removeClientFromCache

lopezlozoya opened the pull request:
#### 10ac38dba49ec8662dd59914a0583084eeb8455b
<pre>
ASAN_TRAP | WTF::HashTable::lookup; WebCore::LegacyRenderSVGResourceClipper::removeClientFromCache; WebCore::SVGResources::removeClientFromCache
<a href="https://bugs.webkit.org/show_bug.cgi?id=288752">https://bugs.webkit.org/show_bug.cgi?id=288752</a>
<a href="https://rdar.apple.com/144407636">rdar://144407636</a>

Reviewed by Ryosuke Niwa.

Change the from UncheckedKeyHashMap to WeakHashMap in m_clipper in LegacyRenderSVGResourceClipper.

* LayoutTests/svg/clip-path/clip-path-remove-null-client-from-cache-expected.txt: Added.
* LayoutTests/svg/clip-path/clip-path-remove-null-client-from-cache.html: Added.
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceClipper.h:

Canonical link: <a href="https://commits.webkit.org/291601@main">https://commits.webkit.org/291601@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9de577703feda4dd28963bcb46a6737cc370f997

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/93203 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/12758 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/2466 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/98202 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/43729 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/95253 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/13037 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/21211 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71238 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28644 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/96205 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/9798 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/84306 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51570 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9493 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/1943 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/43043 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/79782 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/1935 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/100231 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/20255 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/14833 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/80265 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/20507 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/80217 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79574 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19809 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24120 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/1435 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/13370 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/20239 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/25416 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/19926 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/23386 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/21667 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceClipper.h:
```diff
@@ -86,7 +86,7 @@ class LegacyRenderSVGResourceClipper final : public LegacyRenderSVGResourceConta
 
 private:
-    bool selfNeedsClientInvalidation() const override { return (everHadLayout() || m_clipperMap.size()) && selfNeedsLayout(); }
+    bool selfNeedsClientInvalidation() const override { return (everHadLayout() || m_clipperMap.isEmptyIgnoringNullReferences()) && selfNeedsLayout(); }
```
Looks like we should be checking !m_clipperMap.isEmptyIgnoringNullReferences()?

lopezlozoya replied:
Thank you, I'll fix it.

rwlbuis commented on LayoutTests/svg/clip-path/clip-path-remove-null-client-from-cache.html:
```diff
@@ -0,0 +1,26 @@
+        window?.testRunner?.notifyDone();
+    })
+}
+</script>
```
How many test iterations are needed to reproduce the problem? The test in https://github.com/WebKit/WebKit-security/pull/88/files is simpler but usually needs multiple iterations (for example 10 iterations triggers it). Maybe it is possible to change that tests so it does iterations internally (like in above test). Key seems to be removing/destroying the HTML element that references the SVG clip path.

lopezlozoya replied:
Using the test file in this PR, it takes only one execution of it to trigger the crash. Looks like I don't have access to the link you posted here, appears as page not found to me.

rwlbuis replied:
> Using the test file in this PR, it takes only one execution of it to trigger the crash. Looks like I don't have access to the link you posted here, appears as page not found to me.

It is on embargoed branch, maybe you can ask an Apple colleague. At the same time, maybe I can put it on main branch tomorrow.

rniwa commented on LayoutTests/svg/clip-path/clip-path-remove-null-client-from-cache.html:
```diff
@@ -0,0 +1,26 @@
+<body>
+<p>This test passes if it doesn't crash.</p>
+<script>
+if (window.testRunner)
```
This if statement is redundant since we call `ndow?.testRunner?.dumpAsText();` below. Remove?

lopezlozoya replied:
That is right, thanks, I'll fix it.

nikolaszimmermann reviewed (commented):
Hmm it would be better to properly deregister clients to avoid the situation...

lopezlozoya committed:
ASAN_TRAP | WTF::HashTable::lookup; WebCore::LegacyRenderSVGResourceClipper::removeClientFromCache; WebCore::SVGResources::removeClientFromCache
https://bugs.webkit.org/show_bug.cgi?id=288752
rdar://144407636

Reviewed by Ryosuke Niwa.

Change the from UncheckedKeyHashMap to WeakHashMap in m_clipper in LegacyRenderSVGResourceClipper.

* LayoutTests/svg/clip-path/clip-path-remove-null-client-from-cache-expected.txt: Added.
* LayoutTests/svg/clip-path/clip-path-remove-null-client-from-cache.html: Added.
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceClipper.h:

Canonical link: https://commits.webkit.org/291601@main
